src/main/haskell/kore/src/Kore/Step/Substitution.hs

```diff
@@ -52,6 +54,31 @@ import           Kore.Unification.UnifierImpl
                  ( normalizeSubstitutionDuplication )
 import           Kore.Variables.Fresh
 
+
+-- | Normalize the substitution of 'expanded', or return 'bottom' if
+-- normalization fails.
+normalize 
+  :: forall level variable m . 
+     ( level ~ Object
+     , Monad m
+     , MonadCounter m
+     , MetaOrObject level
+     , Hashable variable
+     , FreshVariable variable
+     , SortedVariable variable
+     , OrdMetaOrObject variable
+     , ShowMetaOrObject variable
+     ) 
+  => MetadataTools level StepperAttributes 
+  -> ExpandedPattern level variable 
+  -> m (ExpandedPattern level variable)
+normalize tools r =
+    runExceptT (normalizePredicatedSubstitution tools r)
+        >>= \case
+            Left _ -> return ExpandedPattern.bottom
```

Comment:
Bottom means unification is impossible, Left means we have no clue how to unify, but it might work. So, if you don't know how to unify your lists/list elements (i.e. the error case) you should also return an error (the other option is making a predicate out of that unification).

Please delete `normalize` and use `normalizePredicatedSubstitution` directly.

Reply:
`normalize` was already used in Tom's Map PR, I just moved it to a different spot. So far as I know, my use of it in `Builtin.List.unify` is as correct as the use in `Builtin.Map.unify`

Reply:
Ok, I'll let @ttuegel decide what should be done with this 'normalize' issue.

Reply:
I'm not sure I understand what the issue is. `normalizePredicatedSubstitution` returns `Left` if it doesn't know how to normalize a list of substitutions. That list of substitutions is itself correct and unique for a given unification (sub-) problem.

Reply:
Let me try to explain it this way:

Let phi and psi be two patterns that we can't unify (e.g. `Mock.plain10(Mock.a)` and `Mock.b`).

If someone calls `normalize [x=phi, x=psi]`, and normalize returns bottom, the caller concludes that there is a contradiction in the substitution, so it should give up on whatever it is doing and return bottom (or, if that happened in a `not`, it will return `top`; or something like that). Note that returning bottom is a perfectly valid thing and it should happen for, say, `[x=constr1(a), x=constr2(a)]`.

Now, if `normalize` returns an error, then the caller can postpone normalizing the substitution (by transforming it into a predicate) for a later time, when more information is available. Note that we actually want to do this when, say, postponing a map lookup when the key is a variable, but we probably want to do this in other cases, also.

This difference is also relevant when evaluating functions defined on multiple branches (e.g. signum). For the moment, most likely, we will not evaluate functions on symbolic variables except in some very narrow cases, but, in the future, we may want to know that one of the cases cannot apply to our variables (i.e. bottom), that it might apply (error), or that it can be applied when a certain predicate is satisfiable (normal return value).

Reply:
So, if `normalize` returns bottom instead of Left, or if the error is not propagated above the List/Map unification, we can't make that difference anymore, and we can't postpone evaluating functions.

Reply:
Hmm. I honestly am not sure what the correct behavior is then. I'll have to ask Tom on monday whether his use of `normalize` is correct.

Reply:
I don't understand what `Left` from `normalize` means; is it an error or not? By "error", I mean "unrecoverable", as opposed to "exception".